File: apps/customer-web/lib/cart-totals.ts
/**
 * Client-side cart totals for DISPLAY. Integer-cents math (never floats) mirroring core-api's `money`
 * package. This is a display approximation for the guest cart review — the SERVER computes the
 * authoritative amount at checkout, and the charge is always the server's.
 *
 * 021: there is NO client-side delivery fee any more. Delivery is per-package, geographic, and needs
 * a destination address to price — so it is quoted only at the delivery step (FR-024/SC-010: no order
 * or package ever falls back to a flat/hardcoded fee). The cart shows the item subtotal and says
 * "Delivery calculated at checkout".
 */
import type { GuestCartLine } from "./cart-store"

export function parseCents(amount: string): number {
  const [whole, frac = ""] = amount.replace("-", "").split(".")
  const cents = Number(whole) * 100 + Number((frac + "00").slice(0, 2))
  return amount.startsWith("-") ? -cents : cents
}

export function formatCents(cents: number): string {
  const neg = cents < 0
  const abs = Math.abs(cents)
  return `${neg ? "-" : ""}${Math.floor(abs / 100)}.${String(abs % 100).padStart(2, "0")}`
}

export interface CartTotals {
  /** Σ(unit×qty). The ONLY amount the cart can know before an address is chosen (021). */
  itemSubtotal: string
}

/** Σ(unit×qty). Delivery is NOT included — it is quoted per package at checkout (021 FR-024). */
export function computeCartTotals(lines: readonly GuestCartLine[]): CartTotals {
  const subtotal = lines.reduce((c, l) => c + parseCents(l.unitPriceAmount) * l.quantity, 0)
  return { itemSubtotal: formatCents(subtotal) }
}

File: apps/customer-mobile/shared/src/commonMain/kotlin/com/effyshopping/customer/mobile/features/cart/domain/CartTotals.kt
package com.effyshopping.customer.mobile.features.cart.domain

/**
 * Client-side cart totals for DISPLAY, integer-cents math mirroring core-api's money/pricing.
 *
 * 021: the flat $5 delivery fee is GONE (FR-024). Delivery is now geographic + per-package and is only
 * known once an address is chosen at the delivery step, from the server quote — so the cart shows the
 * ITEM SUBTOTAL ONLY. The server recomputes the authoritative grand total (items + summed per-package
 * fees) at checkout; this is only the guest-cart item estimate.
 */

data class CartTotals(val itemSubtotal: String)

fun parseCents(amount: String): Long {
    val negative = amount.startsWith("-")
    val clean = amount.removePrefix("-")
    val parts = clean.split(".")
    val whole = parts.getOrNull(0)?.toLongOrNull() ?: 0L
    val fracRaw = (parts.getOrNull(1) ?: "") + "00"
    val frac = fracRaw.substring(0, 2).toLongOrNull() ?: 0L
    val cents = whole * 100 + frac
    return if (negative) -cents else cents
}

fun formatCents(cents: Long): String {
    val negative = cents < 0
    val abs = if (negative) -cents else cents
    val whole = abs / 100
    val frac = (abs % 100).toString().padStart(2, '0')
    return "${if (negative) "-" else ""}$whole.$frac"
}

fun computeTotals(lines: List<GuestCartLine>): CartTotals {
    val subtotal = lines.sumOf { parseCents(it.unitPriceAmount) * it.quantity }
    return CartTotals(itemSubtotal = formatCents(subtotal))
}
